src/libsystemd-network/network-internal.c

```diff
@@ -131,6 +131,90 @@ int deserialize_in6_addrs(struct in6_addr **ret, const char *string) {
         return size;
 }
 
+int serialize_dnr(FILE *f, const sd_dns_resolver *dnr, size_t n_dnr, bool *with_leading_space) {
+        int r;
+
+        bool _space = false;
+        if (!with_leading_space)
+                with_leading_space = &_space;
+
+        int n = 0;
+        _cleanup_strv_free_ char **names = NULL;
+        r = sd_dns_resolvers_to_dot_strv(dnr, n_dnr, &names);
+        if (r < 0)
+                return r;
+        if (r > 0)
+                fputstrv(f, names, NULL, with_leading_space);
+        n += r;
+        return n;
+}
+
+/* Deserialized resolvers are assumed to offer DoT service. */
+int deserialize_dnr(sd_dns_resolver **ret, const char *string) {
+        int r;
+
+        assert(ret);
+        assert(string);
+
+        sd_dns_resolver *dnr = NULL;
+        size_t n = 0;
+        CLEANUP_ARRAY(dnr, n, sd_dns_resolver_array_free);
+
+        sd_dns_resolver resolver = {};
+
+        for (;;) {
+                _cleanup_free_ char *word = NULL;
+
+                r = extract_first_word(&string, &word, NULL, 0);
+                if (r < 0)
+                        return r;
+                if (r == 0)
+                        break;
+
+                uint16_t port;
+                int family;
+                _cleanup_free_ union in_addr_union *addr = new(union in_addr_union, 1);
+                _cleanup_free_ char *auth_name = NULL;
+
+                r = in_addr_port_ifindex_name_from_string_auto(word, &family, addr, &port, NULL, &auth_name);
+                if (r < 0)
+                        return r;
+
+                /* coalesce resolvers. Since DoT is assumed, no need to compare transports/dohpath/etc. */
+                bool found = false;
+                FOREACH_ARRAY(res, dnr, n) {
+                        if (family == res->family && streq(auth_name, res->auth_name)) {
+                                if (!GREEDY_REALLOC(res->addrs, res->n_addrs+1))
+                                        return -ENOMEM;
+                                res->addrs[res->n_addrs++] = *addr;
+                                found = true;
+                                break;
+                        }
+                }
+                if (found)
+                        continue;
+
+                resolver = (sd_dns_resolver) {
+                        .priority = n+1, /* not serialized, but this will preserve the order */
+                        .auth_name = TAKE_PTR(auth_name),
+                        .family = family,
+                        .addrs = TAKE_PTR(addr),
+                        .n_addrs = 1,
+                        .transports = SD_DNS_ALPN_DOT,
+                        .port = port,
+                        .dohpath = NULL,
+                        /* list fields are zero-initialized */
+                };
+
+                if (!GREEDY_REALLOC(dnr, n+1))
+                        return -ENOMEM;
+
+                dnr[n++] = TAKE_STRUCT(resolver);
```

Comment:
why write to temp var `resolver` first? you can just use the compound initialize here directly?

Reply:
I think this was written to help the \_cleanup\_ attributes do the right thing in an older version of this function, but you know I'm not sure. Actually, as written it looks buggy right now since if the realloc fails I guess we don't cleanup auth_name...